README
This is the next generation Linux driver for the FSCC line of cards from 
Commtech, Inc. Version 2 is a complete rewrite from the original. The original 
driver was based largely in code that was written for 2.4 era kernels and then 
ported over so there were some spots where it was not exactly using best 
practices for today's kernels. The version 2 driver is essentially complete. 
There are some features and functions that have not been integrated yet, but 
hopefully we will be able to add them soon. WARNING: THIS DRIVER IS NOT YET 
CONSIDERED A STABLE RELEASE. THERE ARE STILL SOME MISSING FUNCTIONS AND 
FEATURES AND ARE LIKELY SOME BUGS.  Use at your own risk.  If you find a bugfix
and would like to submit it to us please email the patch or modified code to
techsupport@commtech-fastcom.com.

I.     Getting the latest driver software
       1) Switching between driver versions
      
II.    Compiling the driver
       1) Enabling debug prints
       2) Specifying non-default header files directory
      
III.   Loading/installed the driver
       1) Installing the driver
       2) Enabling hot plugging support
       3) Disabling DMA support

V.     Changing a port's register values
       1) Using sysfs from the command line
       2) Using the ioctl call from within your application code
       3) Using the Python API from within your application code
       4) Using the config.h file for modifying the boot defaults

VI.    Reading a port's register values
       1) Using sysfs from the command line
       2) Using the ioctl call from within your application code
       3) Using the Python API from within your application code
      
VII.   Asynchronous communication

VIII.  Setting the clock frequency
       1) Using the ioctl call from within your application code

IX.    Operating the driver
       1) Writing data in C code
       2) Writing data from the command line
       3) Writing data using the Python API
       4) Reading data in C code
       5) Reading data from the command line
       6) Reading data using the Python API
       
X.     Viewing/setting frame status data
       1) Using sysfs from the command line
       2) Using the ioctl call from within yzour application code
       3) Using the config.h file for modifying the boot defaults
       
XI.    Viewing/setting memory constraints
       1) Using sysfs from the command line
       2) Using the ioctl call from within yzour application code
       3) Using the config.h file for modifying the boot defaults

XII.   Purging data out of driver
       1) Using sysfs from the command line
       2) Using the ioctl call from within your application code
       2) Using the Python API from within your application code
       
XIII.  Migrating from the 1.x driver to the 2.x driver
       1) Setting register values
       2) Getting register values
       3) Purging transmit and receive data
       4) Getting the frame status
       5) Changing the clock frequency
       
XIV.   FAQ
       1) Ports are not created even though the driver has loaded
       2) What does poll() and select() base their information on
       3) Executing a purge without a clock puts the card in a broken state
       4) Receiving the error message 'Couldn't register serial port'
       5) Not seeing card in sysfs


I. Getting the latest driver software
-------------------------------------------------------------------------------
The source code for the FSCC driver is hosted on bitbucket code hosting. To check 
out the latest code you will need Mercurial and to run the following in a 
terminal.

   > hg clone https://bitbucket.org/fastcom/fscc-linux fscc-linux

NOTE: We prefer you use the above method for downloading the driver (because it 
      is the easiest way to stay up to date) but you can also download a 
      tarball from the following location:
   
      http://hg.commtech-fastcom.com/fscc-linux/downloads 

1) Now that you have the latest code checked out you will more than likely want 
   to switch to a stable version within the code directory. To do this browse 
   the various tags for one you would like to switch to. Version v2.0.3 is only
   listed here as an example.

   > hg tags
   > hg update v2.0.3
   
   
II. Compiling the driver
-------------------------------------------------------------------------------
Compiling the driver is relatively simple assuming you have all of the 
required dependencies. Typically you will need gcc, make and your kernel's
header files. After assembling all of these things you can build the driver by
simply running the make command from within the source code directory.

   > cd fscc-linux/
   > make

1) If you would like to enable debug prints within the driver you need to add 
   the DEBUG option while building the driver.

   > make DEBUG=1

   Once debugging is enabled you will find extra kernel prints in the
   /var/log/messages and /var/log/debug log files.
   
2) If the kernel header files you would like to build against are not in the
   default location '/lib/modules/$(shell uname -r)/build' then you can specify
   the location with the KDIR option while building the driver.

   > make KDIR="/location/to/kernel_headers/"


III. Loading/installing the driver
-------------------------------------------------------------------------------
Assuming the driver has been successfully built in the previous step you are
now ready to load the driver so you can begin using it. To do this you insert
the driver's kernel object file (fscc.ko) into the kernel.

   > insmod fscc.ko

NOTE: You will more than likely need administrator privileges for this and
      the following commands.

1) If you would like the driver to load automatically at boot use the included
   installer.
   
   > make install
   
   This will also install the header (.h) files, python library and command
   line program for using the card.

   NOTE: The driver can be removed with: make uninstall
      
2) By default if there are no cards present when the driver is loaded the
   insmod will fail with a 'No such device' error. To allow the driver to load
   even if no cards are present turn on the 'hot_plug' option.
   
   > insmod fscc.ko hot_plug=1
   
   If no cards are present and the hot_plug option is not enabled (default) you
   will see this.
   
   > insmod fscc.ko hot_plug=0
   insmod: error inserting 'fscc.ko': -1 No such device
   
3) If you would like to disable DMA for cards (SuperFSCC* line) that have that
   capability you can do so by using the 'force_fifo' option. This option will
   disable DMA and allow you to operate in the FIFO based mode.
   
   > insmod fscc.ko force_fifo=1
   
NOTE: All driver load time options can be set in your modprobe.conf file for
      using upon system boot.


V. Changing a port's register values
-------------------------------------------------------------------------------
The FSCC driver is a swiss army knife of sorts with communication. It can
handle many different situations if configured correctly. Typically to 
configure it to handle your specific situation you need to modify the card's
register values.

There are multiple ways of modifying the card's registers varying from simply
modifying a sysfs file to an ioctl call within code. Here are a few ways of
doing this.

NOTE: For a listing of all of the configuration options please see the manual.

1) Echo the register value in hexadecimal form to the sysfs file with the same
   name of the register you would like to modify.

   > echo 030000ff > /sys/class/fscc/fscc0/registers/bgr
   
2) Use the FSCC_SET_REGISTERS ioctl to set the values of any registers you
   need to modify from within C code. This ioctl can be found within 
   <fscc/fscc.h> after you install the header files (see section IV).
   
   struct fscc_registers regs;
   
   FSCC_REGISTERS_INIT(regs);
   
   regs.BGR = 0x030000ff;
   regs.FCR = 0x00000000;
   
   ioctl(port_fd, FSCC_SET_REGISTERS, &regs);
   
   NOTE: A complete example of how to do this can be found in the file 
         fscc-linux/examples/c/set-registers.c.
         
3) Use the Python API to easily set the values of any registers you need to
   modify from within Python code.

   >>> port.registers.BGR = 0x030000ff
   >>> port.registers.FCR = 0x00000000
   
   NOTE: A complete example of how to do this can be found in the file 
         fscc-linux/examples/python/set-registers.py.
   
4) Modify the '#define DEFAULT_[BGR|CCR0|...]_VALUE 0x030000ff' lines within 
   the config.h file to be whatever you would like the card to boot up as. You 
   will need to recompile the driver after doing this.
   
   NOTE: This will set all ports to this at driver boot up. It is a driver wide
         setting.
   

VI. Reading a port's register values
-------------------------------------------------------------------------------
There are multiple ways of reading the card's registers varying from simply
modifying a sysfs file to an ioctl call within code. Here are a few ways of
doing this.

1) Cat the register value in hexadecimal form from the sysfs file with the same
   name of the register you would like to view.
   
   > cat /sys/class/fscc/fscc0/registers/bgr
   00000001
   
2) Use the FSCC_GET_REGISTERS ioctl to get the values of any registers you
   need to read from within code. This ioctl can be found within 
   <fscc/fscc.h> after you install the header files (see section IV).
   
   struct fscc_registers regs;
   
   FSCC_REGISTERS_INIT(regs);
   
   regs.BGR = FSCC_UPDATE_VALUE;
   regs.FCR = FSCC_UPDATE_VALUE;
   
   ioctl(port_fd, FSCC_GET_REGISTERS, &regs);
   
   At this point 'regs.BGR' and 'regs.FCR' would be set to their respective 
   values.
   
   NOTE: A complete example of how to do this can be found in the file 
         fscc-linux/examples/c/get-registers.c.
         
3) Use the Python API to easily get the values of any registers you need to
   read from within Python code.

   >>> "0x%08x" % port.registers.BGR
   "0x00000000"
   
   >>> "0x%08x" % port.registers.FCR
   "0x00000000"
   
   NOTE: A complete example of how to do this can be found in the file 
         fscc-linux/examples/python/get-registers.py.


VII. Asynchronous communication
-------------------------------------------------------------------------------
The FSCC driver lets the built in Linux serial driver handle asynchronous
communication for our UARTs. The Linux serial driver is highly tested and 
likely more stable than anything we could produce in any reasonably amount of 
time.

Prior to and after loading the FSCC driver (see section III) there are a few 
steps needed to get the card ready for asynchronous communication.

1) Some Linux distributions have the default number of serial ports that are
   available at boot set to a small number (usually 4). The first four serial
   ports are reserved so you will need to change this value to something larger
   to be able to configure more serial ports.
   
   There are a couple ways of doing this. The easiest method is by appending 
   '8250.nr_uarts=x' to your grub boot line. Something like this:
   
   kernel /boot/vmlinuz-2.6.20-15-generic ro quiet splash 8250.nr_uarts=16
   
   This can be done temporarily by pressing 'e' at the grub menu during boot or 
   by permanently modifying this value which is grub version specific. To do
   this please search google for one of the numerous guides on the subject.
   
   Another method is by editing the .config file of you kernel before compiling
   it to allow for more serial ports. This is not preferred because you will
   need to recompile the kernel for it to take effect. The line you need to
   change in the .config file is SERIAL_8250_RUNTIME_UARTS.

2) Load the FSCC driver. This will handle registering our UARTs with the serial
   driver (see section III). Our UART's will now appear as ttyS nodes in the 
   /dev/ directory.

3) By default the FSCC driver boots up in synchronous communication mode. To
   switch to the asynchronous mode you must modify the FSCC card's FCR register 
   to allow for asynchronous communication. There are multiple ways of doing 
   this (see section V). Possibly the simplest method is using sysfs and the
   command line.
   
   > echo 03000000 > /sys/class/fscc/fscc0/registers/fcr


VIII. Setting the clock frequency
-------------------------------------------------------------------------------
The FSCC device has a programmable clock that can be set anywhere from
20 kbit to 200 Mbit. However, this is not the fully operational range an
FSCC port.  This is the range that the onboard clock can be set to.

Using one of the synchronous modes you can only receive data consistently 
up to 30 Mbit (when you are using a external clock). If you are transmitting 
data using an internal clock you can safely go up 50 Mbit.

1) Use the FSCC_SET_CLOCK_BITS ioctl to set the frequency from within code.
   
   /* 10 MHz */
   unsigned char clock_bits[20] = {0x01, 0xa0, 0x04, 0x00, 0x00, 0x00, 0x00,
	                               0x00, 0x00, 0x00, 0x00, 0x9a, 0x4a, 0x41,
	                               0x01, 0x84, 0x01, 0xff, 0xff, 0xff};
	                                
   ioctl(port_fd, FSCC_SET_CLOCK_BITS, &clock_bits);
   
   NOTE: A complete example of how to do this along with how to calculate
         these clock bits can be found in the file 
         fscc-linux/examples/c/set-clock-bits.c.


IX.  Operating the driver
-------------------------------------------------------------------------------
The FSCC driver typically (but not always) works in "frames". This means that 
data typically is packaged together in permanent chunks. If the card received
two frames of data prior to you retrieving the data you will only get one chunk
of data back at a time when you interface with the driver to get the data.

There are multiple ways of reading/writing data to/from the card. Listed below
are only the most common.

1) Writing data will typically be done within C code using the write()
   function found within the <unistd.h> file. You can find a more in depth
   example named sync-write.c within the examples/c directory. You will need
   to look elsewhere for a more in depth explanation of the parameters and 
   return values.
   
   bytes_written = write(port_fd, data, sizeof(data));
   
   In in addition to the standard errors that the write() function returns
   there are a couple errors specific to the FSCC you might run into.
   
   -EOPNOTSUPP: If trying to use a FSCC port with the /dev/fscc nodes while
                in asynchronous communication mode (should be using /dev/ttyS
                nodes).
                
   -ETIMEDOUT: If trying to use a FSCC port without a transmit clock present.
               This check can be turned off with the 'ignore_timeout' command
               line parameter.
   
   -ENOBUFS: If the count parameter passed into the write() function is larger
             than the output cap. If the count parameter is less than the
             output cap but the amount out of output space isn't enough the 
             driver will block instead of returning this error.
   
   NOTE: A complete example of how to do this can be found in the file 
         fscc-linux/examples/c/sync-write.c.
   
2) A simpler but not nearly as useful way of sending data out of a FSCC port is
   by using the built-in linux program echo. This is mainly useful while
   debugging.
   
   echo "Hello world!" > /dev/fscc0
   
3) The Python API has a built-in method for writing data to the card.

   >>> port.write('Hello world!'.encode())
    
   The same errors returned from the C code above can be found while using the
   Python API. The only difference is instead of returning an error they throw
   an IOError exception with the appropriate errno value set.
   
   NOTE: A complete example of how to do this can be found in the file 
         fscc-linux/examples/python/write.py.
   
4) Reading data will typically be done within C code using the read() function
   found within the <unistd.h> file. You can find a more in depth example named
   sync-read.c within the examples/c directory. You will need to look elsewhere
   for a more in depth explanation of the parameters and return values.
      
   bytes_read = read(port_fd, data, sizeof(data));
   
   The length argument of the read() function means different things depending 
   on the mode you are using. 
   
   In a frame based mode the length argument specifies the maximum frame size 
   to return. If the next queued frame is larger than the size you specified 
   the error -ENOBUFS is returned and the data will remain waiting for a read() 
   of a larger value. If a read() length is specified that is larger than the
   length of multiple frames in queue you will still only receive one frame per 
   read() call.
   
   In streaming mode (no frame termination) the length argument specifies the 
   maximum amount of data to return. If there is 100 bytes of streaming data 
   in the card and you read() with a length of 50, you will receive 50 bytes. 
   If you were to do a read() of 200 bytes you will receive the 100 bytes 
   available.
   
   Frame based data and streaming data are kept separate within the driver.
   To understand what this means first imagine this scenario. You are in a 
   frame based mode and receive a couple of frames. You then switch to 
   streaming mode and receive a stream of data. When calling read() you will 
   receive the the streaming data until you switch back into a frame based 
   mode then do a read().
   
   In in addition to the standard errors that the read() function returns
   there are a couple errors specific to the FSCC you might run into.
   
   -ENOBUFS: If the size parameter passed into the read() function is smaller 
             than the next frame (in a frame based mode).
   
   -EOPNOTSUPP: If trying to using a FSCC port using the /dev/fscc nodes while
                in asynchronous communication mode (should be using /dev/ttyS
                nodes).
               
   
   NOTE: A complete example of how to do this can be found in the file 
         fscc-linux/examples/c/sync-read.c.
   
5) A simpler but not nearly as useful way of reading data out of a FSCC port is
   by using the built-in linux program cat. This is mainly useful while
   debugging.
   
   cat /dev/fscc0
   
6) The Python API has a built-in method for read data from the card.

   >>> port.read(4096)
    
   The same errors returned from the C code above can be found while using the
   Python API. The only difference is instead of returning an error they throw
   an IOError exception with the appropriate errno value set.
   
   NOTE: A complete example of how to do this can be found in the file 
         fscc-linux/examples/python/read.py.
   
NOTE: fread/fwrite are not allowable ways of interfacing with the card. They
      make assumptions about the data that doesn't hold up with our "frame"
      perspective. This means Python functions like file.write() and
      file.read() will not work because they are wrappers around fread/fwrite.
      You will instead need to use file streams which act more like read/write.
      
        
X. Viewing/setting frame status data
-------------------------------------------------------------------------------
It is a good idea to pay attention to the status of each frame. For example if 
you want to see if the frame's CRC check succeeded or failed. 
   
The way the FSCC reports this data to you is by appending two additional bytes 
to each frame you read from the card if you opt in to see this data. There are 
a couple of ways of enabling this additional data.
   
1) The simplest method of doing this is by writing a value of 1 to the
   append_status file in sysfs from the command line.
   
   echo 1 > /sys/class/fscc/fscc0/settings/append_status
   
   You can also check if append status is enabled from the command line.
   
   cat /sys/class/fscc/fscc0/settings/append_status
   
2) Another way of doing this is by using the FSCC_ENABLE_APPEND_STATUS ioctl
   from within your C code.
   
   ioctl(port_fd, FSCC_ENABLE_APPEND_STATUS);
   
   NOTE: A complete example of how to do this can be found in the file 
         fscc-linux/examples/c/append-status.c.
   
3) Modify the '#define DEFAULT_APPEND_STATUS_VALUE 1' lines within the config.h 
   file to be whatever you would like the card to boot up as. You will need to 
   recompile the driver after doing this.
   
   NOTE: This will set all ports to this at driver boot up. It is a driver wide
         setting.
         
         
XI. Viewing/setting memory constraints
-------------------------------------------------------------------------------
For systems with limited memory available to them there is safety checks in 
place to prevent spurious incoming data from overrunning your system. Each port
has an option for setting it's input and output memory cap.

There are multiple ways of setting this value.
   
1) The simplest method of doing this is by writing a value to the
   {input, output}_memory_cap file in sysfs from the command line.
   
   echo 1000000 > /sys/class/fscc/fscc0/settings/input_memory_cap
   
   You can also check the current memory cap value from the command line.
   
   cat /sys/class/fscc/fscc0/settings/input_memory_cap
   
2) Use the FSCC_SET_MEMORY_CAP ioctl to set the memory cap values from within C 
   code. This ioctl can be found within <fscc/fscc.h> after you install the 
   header files (see section IV).
   
   struct fscc_memory_cap memory_cap;
   
   FSCC_MEMORY_CAP_INIT(memory_cap);
   
   memory_cap.input = 5000000;
   memory_cap.output = 10000;
   
   ioctl(port_fd, FSCC_SET_MEMORY_CAP, &memory_cap);
   
   NOTE: You can set only 1 of the 2 values by running the FSCC_MEMORY_CAP_INIT
         macro on the fscc_memory_cap struct then setting only 1 of the values
         in the structure. The FSCC_MEMORY_CAP_INIT structure initializes both
         values to -1 which will be ignored in the driver.
   
   NOTE: A complete example of how to do this can be found in the file 
         fscc-linux/examples/c/memory-cap.c.
   
3) Modify the '#define DEFAULT_{INPUT, OUTPUT}_MEMORY_CAP 100000' lines within 
   the config.h file to be whatever you would like the card to boot up as. You
   will need to recompile the driver after doing this.
   
   NOTE: This will set all ports to this at driver boot up. It is a driver wide
         setting.

XII. Purging data out of driver
-------------------------------------------------------------------------------
Between the hardware FIFO and the driver's software buffers there are multiple
places data could be at excluding your application code. If you ever need to 
clear this data out and start out fresh there are a couple ways of doing this.
   
1) The simplest method of doing this is by writing a value of 1 to the
   purge_tx and purge_rx file in sysfs from the command line.
   
   echo 1 > /sys/class/fscc/fscc0/commands/purge_tx
   echo 1 > /sys/class/fscc/fscc0/commands/purge_rx
   
2) Another way of doing this is by using the FSCC_PURGE_TX and FSCC_PURGE_TX
   ioctl's from within your C code.
   
   ioctl(port_fd, FSCC_PURGE_TX);
   ioctl(port_fd, FSCC_PURGE_RX);
   
   NOTE: A complete example of how to do this can be found in the files 
         fscc-linux/examples/c/purge_tx.c and purge_rx.c.
   
   In in addition to the standard errors that the ioctl() function returns
   there is an error specific to the FSCC you might run into.
                
   -ETIMEDOUT: If trying to use a FSCC port without a transmit clock present.
               This check can be turned off with the 'ignore_timeout' command
               line parameter.
         
3) The Python API makes this easy by calling a couple built-in methods.

   >>> port.purge_rx()
   >>> port.purge_tx()
   
   
XIII. Migrating from the 1.x driver to the 2.x driver
-------------------------------------------------------------------------------
There are multiple benefits of using the 2.x driver: accurate posix error
codes, intuitive ioctl calls, backend support for multiple languages (Python,
C#) and dynamic memory management are some.

The 1.x driver and the 2.x driver are very similar so porting from one to the
other should be rather painless.

NOTE: All ioctl values have changed even if their new names match their old
      names. This means even if you use an ioctl with an identical name, it
      will not work correctly.

1) Setting register values was split into two different ioctl's in the 1.x
   driver, setting all the registers at once and one at a time. In the 2.x 
   driver these two scenarios have been combined into one ioctl.
   
   Change the following ioctl's to the current ioctl FSCC_SET_REGISTERS (see 
   section V).
   
   FSCC_WRITE_REGISTER (setting a single register at a time)
   FSCC_INIT (setting all registers at a time)
   
2) Getting register values was limited to one at a time in the 1.x driver. In
   the 2.x driver it has been made more convenient to read multiple register
   values.
   
   Change the following ioctl to the current ioctl FSCC_GET_REGISTERS (see
   section VI).
   
   FSCC_READ_REGISTER (reading a single register at a time)
   
3) Purging transmit and receive data has not changed. Continue using
   FSCC_PURGE_TX and FSCC_PURGE_RX. 
   
   For more information on these ioctl's see section XII.
   
4) Getting the frame status has now been designed to be configurable. In the 
   1.x driver you would always have the frame status appended to your data on a
   read. In the 2.x driver this can be toggled, and defaults to not appending
   the status to the data. 
   
   For more information on the ioctl's to toggle this feature see section X.
   
5) Changing the clock frequency is basically the same but the data structure 
   and ioctl name are different.
   
   Change the following ioctl to the current ioctl FSCC_SET_CLOCK_BITS (see
   section VIII).
   
   FSCC_SET_FREQ (setting the clock frequency)
   
   In the 1.x driver you passed in a structure composed of both the desired

ChangeLog
version 2.4.8 (12/09/2013)
--------------------------------------------------------------------------------    
    * Updated SerialFC to v2.2.0
      (https://github.com/commtech/serialfc-linux/blob/master/ChangeLog.txt)
    * Increased the default timeout duration to 50 cycles
    * Added early support for getnstimeofday()
    * Fixed setting the clock rate on 232 cards
    * Fixed a buf where DMA frames were getting lost
    * Various small fixes


version 2.4.7 (11/14/2013)
--------------------------------------------------------------------------------
    * Fixed a bug that prevented small frame sizes from being calculated 
      correctly when approaching the memory limit


version 2.4.6 (11/5/2013)
--------------------------------------------------------------------------------
    * Fixed a bug where outgoing frames wouldn't get cleared fast enough


version 2.4.5 (10/7/2013)
--------------------------------------------------------------------------------
    * Fixed a bug where input stream data wasn't being calculated in memory
      usage
    * Fixed tx_modifiers ioctl to accept raw integers


version 2.4.4 (09/25/2013)
--------------------------------------------------------------------------------
    * Fixed a regression where card's weren't detected on kernels < 3.8


version 2.4.3 (09/25/2013)
--------------------------------------------------------------------------------
	* Performance improvements when transmitting and receiving frames
	* Fix bug preventing tx_modifiers from being set correctly
	* Remove old CLI from project
	* Add links to external libraries
	* Fixed a bug where card won't be recognized if serialfc driver is loaded
	  first
	* Remove hot_plug option


version 2.4.2 (08/28/2013)
--------------------------------------------------------------------------------
    * Fixed a bug that would allow you to return more data than your buffer
      size (introduced in 2.4.0)


version 2.4.1 (08/28/2013)
--------------------------------------------------------------------------------
    * Stability improvements


version 2.4.0 (08/26/2013)
--------------------------------------------------------------------------------
    * Added support for timestamps


version 2.3.0 (08/14/2013)
--------------------------------------------------------------------------------
    * Now use serialfc driver for asynchronous data
    * Optimize write handling
    * Added sysfs support for returning multiple frames


version 2.2.4 (06/20/2013)
--------------------------------------------------------------------------------
    * Added support for returning multiple frames per read call
    * Added support for new card ID's


version 2.2.3 (06/13/2013)
--------------------------------------------------------------------------------
    * Fixed a bug where having NTB set wouldn't put you in a frame based mode
    * Fixed a compiler issue on older kernels
    * Several small performance improvements when handling incoming frames


version 2.2.2 (03/25/2013)
--------------------------------------------------------------------------------
    * Performance improvements
    * Fixed a bug where asynchronous baud rates were off
    * Fixed compiler errors introduced in linux 3.8
    * Fixed random crash in framing based modes


version 2.2.1 (11/16/2012)
--------------------------------------------------------------------------------
    * Fixed a bug where X-Sync mode wasn't considered stream data if there
      wasn't any framing
    * Fixed a data loss bug in transparent mode
    * Improved debug prints


version 2.2.0 (10/09/2012)
--------------------------------------------------------------------------------
    * Added install option to Makefile
    * Improve debug prints
    * Fixed typos
    * Include the gui, cli and pyfscc into the project
    * Added some 'get' ioctls
    * Added additional examples
    * Rename flush{tx,rx} to purge{tx,rx}
    * Update licenses
    * Added fscc.rules file
    * Fixed a bug causing the number of frames to be incorrect
    * Added driver specific registers to sysfs when in debug mode
    * Fixed an issue where a frame was left in the FIFO
    * Optimize the memory consumtion while in streaming mode
    * Fixes setting the clock on older green cards.


version 2.1.3 (02/06/2012)
--------------------------------------------------------------------------------
    * Fixed typos
    * Remove sudo from install_tools
    * Remove make all
    * Added default transmit_modifiers
    * Added tx_modifiers and ignore_timeout to sysfs
    * Rearrange initialization code
    * Use sema_init exclusively
    * Clean up code
    * See changesets for full list of changes


version 2.1.2 (09/14/2011)
--------------------------------------------------------------------------------
    * init_MUTEX(x) removed in 2.6.37, I replaced it with sema_init(x,1)


version 2.1.1 (07/19/2011)
--------------------------------------------------------------------------------
    * Fixed bug that caused the SuperFSCC/4-LVDS to not install correctly


version 2.1.0 (05/02/2011)
--------------------------------------------------------------------------------
    * Fixed comments in code
    * Added make help target
    * Migrate from google code to bitbucket
    * Added code and examples for various transmit modes
    * Fixed dma insmod parameter
    * Added SuperFSCC-LVDS and SuperFSCC/104 to the driver
    * See changesets for full list of changes


version 2.0.5 (02/17/2011)
--------------------------------------------------------------------------------
    * Added a force_fifo option that prevents DMA operation
    * Added a simple example showing how to use the select() function to
      determine how much data is available in the card


version 2.0.4 (02/09/2011)
--------------------------------------------------------------------------------
Bugfix release targeting old DMA APIs.

    * Fixed a build error with kernels prior to 2.6.27 that used an older dma

src/fscc.h
/*
	Copyright (C) 2013 Commtech, Inc.

	This file is part of fscc-linux.

	fscc-linux is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	fscc-linux is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with fscc-linux.	If not, see <http://www.gnu.org/licenses/>.

*/

#ifndef FSCC_H
#define FSCC_H

#include <linux/sched.h> /* wait_queue_head_t */
#include <linux/fs.h> /* struct indode on <= 2.6.19 */

#define FSCC_REGISTERS_INIT(registers) memset(&registers, -1, sizeof(registers))
#define FSCC_MEMORY_CAP_INIT(memory_cap) memset(&memory_cap, -1, sizeof(memory_cap))
#define FSCC_UPDATE_VALUE -2

#define FSCC_IOCTL_MAGIC 0x18

#define FSCC_GET_REGISTERS _IOR(FSCC_IOCTL_MAGIC, 0, struct fscc_registers *)
#define FSCC_SET_REGISTERS _IOW(FSCC_IOCTL_MAGIC, 1, const struct fscc_registers *)

#define FSCC_PURGE_TX _IO(FSCC_IOCTL_MAGIC, 2)
#define FSCC_PURGE_RX _IO(FSCC_IOCTL_MAGIC, 3)

#define FSCC_ENABLE_APPEND_STATUS _IO(FSCC_IOCTL_MAGIC, 4)
#define FSCC_DISABLE_APPEND_STATUS _IO(FSCC_IOCTL_MAGIC, 5)
#define FSCC_GET_APPEND_STATUS _IOR(FSCC_IOCTL_MAGIC, 13, unsigned *)

#define FSCC_SET_MEMORY_CAP _IOW(FSCC_IOCTL_MAGIC, 6, struct fscc_memory_cap *)
#define FSCC_GET_MEMORY_CAP _IOR(FSCC_IOCTL_MAGIC, 7, struct fscc_memory_cap *)

#define FSCC_SET_CLOCK_BITS _IOW(FSCC_IOCTL_MAGIC, 8, const unsigned char[20])

#define FSCC_ENABLE_IGNORE_TIMEOUT _IO(FSCC_IOCTL_MAGIC, 10)
#define FSCC_DISABLE_IGNORE_TIMEOUT _IO(FSCC_IOCTL_MAGIC, 11)
#define FSCC_GET_IGNORE_TIMEOUT _IOR(FSCC_IOCTL_MAGIC, 15, unsigned *)

#define FSCC_SET_TX_MODIFIERS _IOW(FSCC_IOCTL_MAGIC, 12, const unsigned)
#define FSCC_GET_TX_MODIFIERS _IOR(FSCC_IOCTL_MAGIC, 14, unsigned *)

#define FSCC_ENABLE_RX_MULTIPLE _IO(FSCC_IOCTL_MAGIC, 16)
#define FSCC_DISABLE_RX_MULTIPLE _IO(FSCC_IOCTL_MAGIC, 17)
#define FSCC_GET_RX_MULTIPLE _IOR(FSCC_IOCTL_MAGIC, 18, unsigned *)

#define FSCC_ENABLE_APPEND_TIMESTAMP _IO(FSCC_IOCTL_MAGIC, 19)
#define FSCC_DISABLE_APPEND_TIMESTAMP _IO(FSCC_IOCTL_MAGIC, 20)
#define FSCC_GET_APPEND_TIMESTAMP _IOR(FSCC_IOCTL_MAGIC, 21, unsigned *)


enum transmit_modifiers { XF=0, XREP=1, TXT=2, TXEXT=4 };
typedef __s64 fscc_register;

struct fscc_registers {
	/* BAR 0 */
	fscc_register reserved1[2];

	fscc_register FIFOT;

	fscc_register reserved2[2];

	fscc_register CMDR;
	fscc_register STAR; /* Read-only */
	fscc_register CCR0;
	fscc_register CCR1;
	fscc_register CCR2;
	fscc_register BGR;
	fscc_register SSR;
	fscc_register SMR;
	fscc_register TSR;
	fscc_register TMR;
	fscc_register RAR;
	fscc_register RAMR;
	fscc_register PPR;
	fscc_register TCR;
	fscc_register VSTR; /* Read-only */

	fscc_register reserved3[1];

	fscc_register IMR;
	fscc_register DPLLR;

	/* BAR 2 */
	fscc_register FCR;
};

struct fscc_memory_cap {
	int input;
	int output;
};

extern struct list_head fscc_cards;

#define COMMTECH_VENDOR_ID 0x18f7

#define FSCC_ID 0x000f
#define SFSCC_ID 0x0014
#define SFSCC_104_LVDS_ID 0x0015
#define FSCC_232_ID 0x0016
#define SFSCC_104_UA_ID 0x0017
#define SFSCC_4_UA_ID 0x0018
#define SFSCC_UA_ID 0x0019
#define SFSCC_LVDS_ID 0x001a
#define FSCC_4_UA_ID 0x001b
#define SFSCC_4_LVDS_ID 0x001c
#define FSCC_UA_ID 0x001d
#define SFSCCe_4_ID 0x001e
#define SFSCC_4_CPCI_ID 0x001f
#define SFSCC_4_UA_CPCI_ID 0x0023
#define SFSCC_4_UA_LVDS_ID 0x0025
#define SFSCC_UA_LVDS_ID 0x0026

#define STATUS_LENGTH 2

#endif

